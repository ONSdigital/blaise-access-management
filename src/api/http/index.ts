import axios from "axios";

export * from "./users";
export * from "./roles";

export function handleAxiosError(error: unknown): { status: number; message: string, error: unknown } {
    if (axios.isAxiosError(error)) {
        return {
            status: error.response?.status || 500,
            message: error.response?.data.message || "Axios error occurred with no specific message",
            error: error
        };
    } else {
        return {
            status: 500,
            message: "An unknown error occurred, please raise a ServiceNow call if this continues: " + error,
            error: error
        };
    }
}