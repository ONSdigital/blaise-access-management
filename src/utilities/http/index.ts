export * from "./users";
export * from "./roles";
export * from "./auth";
