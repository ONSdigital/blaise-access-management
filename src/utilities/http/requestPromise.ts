import {AuthManager} from "blaise-login-react-client";

type PromiseResponse = [number, any];

function requestPromiseJson(method: string, url: string, body: any = null, headers: any = {}): Promise<PromiseResponse> {
    const authManager = new AuthManager();

    return new Promise((resolve: (object: PromiseResponse) => void, reject: (error: string) => void) => {
        fetch(url, {
            "method": method,
            "body": body,
            "headers": Object.assign({}, headers, authManager.authHeader())
        })
            .then(async response => {
                response.json().then(
                    data => (resolve([response.status, data]))
                ).catch((error) => {
                    resolve([response.status, "Read JSON failed"]);
                });
            })
            .catch(err => {
                console.error(err);
                reject(err);
            });
    });
}

type PromiseResponseList = [boolean, []];

function requestPromiseJsonList(method: string, url: string, body: any = null): Promise<PromiseResponseList> {
    const authManager = new AuthManager();
    return new Promise((resolve: (object: PromiseResponseList) => void, reject: (error: string) => void) => {
        fetch(url, {
            "method": method,
            "body": body,
            "headers": authManager.authHeader()
        })
            .then(async response => {
                response.json().then(
                    data => {
                        if (response.status === 200) {
                            if (!Array.isArray(data)) {
                                resolve([false, []]);
                            }
                            resolve([true, data]);
                        } else if (response.status === 404) {
                            resolve([true, data]);
                        } else {
                            resolve([false, []]);
                        }
                    }
                ).catch((error) => {
                    console.error(error);
                    resolve([false, []]);
                });
            })
            .catch(err => {
                reject(err);
            });
    });
}

export {requestPromiseJson, requestPromiseJsonList};
