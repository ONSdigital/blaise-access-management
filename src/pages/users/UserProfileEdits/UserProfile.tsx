import React, { useCallback, useEffect, useState } from "react";
import { Link, useLocation, useParams } from "react-router-dom";
import Breadcrumbs from "../../../Components/Breadcrumbs";
import ProfileTable from "./ProfileTable";
import { getUser } from "../../../api/http";
import { ONSPanel, ONSLoadingPanel, ONSErrorPanel } from "blaise-design-system-react-components";
import { GetUserResponse } from "../../../Interfaces/usersPage";
import UserSignInErrorPanel from "../../../Components/UserSignInErrorPanel";

export default function UserProfile() {
    const { user: viewedUsername } = useParams();
    const { state } = useLocation();
    const { currentUser, updatedPanel } = state || { currentUser: null, updatedPanel: null };
    const [viewedUserDetails, setViewedUserDetails] = useState<GetUserResponse | null>(null);
    const [loading, setLoading] = useState<boolean>(true);
    const [error, setError] = useState<string>("");

    const fetchUserDetails = async () => {
        setLoading(true);
        try {
            const data = await getUser(viewedUsername as string);
            setViewedUserDetails(data);
            setError("");
        } catch (err) {
            setError("Unable to load user details, please try again. If this continues, please the contact service desk.");
            setViewedUserDetails(null);
        } finally {
            setLoading(false);
        }
    };

    useEffect(() => {
        if (viewedUsername) {
            fetchUserDetails();
        }
    }, [viewedUsername]);

    if (!currentUser) {
        return <UserSignInErrorPanel />;
    }

    return (
        <>
            <Breadcrumbs BreadcrumbList={
                [
                    { link: "/", title: "Home" },
                    { link: "/users", title: "Manage users" }
                ]
            } />
            {updatedPanel && updatedPanel.visible ? (
                <ONSPanel status={updatedPanel.status}>
                    <div className="ons-panel__body">{updatedPanel.message}</div>
                </ONSPanel>)
                : null }
            {error && <ONSPanel status={"error"}>
                <div className="ons-panel__body">{error}</div>
            </ONSPanel>}
            {loading ? (
                <ONSLoadingPanel />
            ) : (
                viewedUserDetails && <ProfileTable currentUser={currentUser} viewedUserDetails={viewedUserDetails} />
            )}
        </>
    );
}