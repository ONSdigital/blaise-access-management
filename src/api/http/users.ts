import axios from "axios";
import { requestPromiseJson, requestPromiseJsonList } from "./requestPromise";
import { User, NewUser } from "blaise-api-node-client";
import { AuthManager } from "blaise-login-react/blaise-login-react-client";
import { GetUserResponse, GetUsersListResponse, PatchUserRoleResponse } from "../../Interfaces/usersPage";
import { handleAxiosError } from ".";

async function getAllUsers(): Promise<GetUsersListResponse> {
    try {
        const url = "/api/users";
        const [success, data] = await requestPromiseJsonList<User>("GET", url);
        return [success, data] as GetUsersListResponse;
    } catch (error) {
        return [false, []] as GetUsersListResponse;
    }
}

// NOTE: Catch block is different due to the use of AsyncContent and AsyncRequest
async function getUser(user: string): Promise<GetUserResponse> {
    try {
        const authManager = new AuthManager();
        const url = `/api/users/${user}`;
        const headers = authManager.authHeader();

        const res = await axios.get(url, { headers });

        return {
            status: res.status,
            message: res.data.message,
            data: res.data.data
        };

    } catch (error) {
        const errorObj = handleAxiosError(error);
        console.error(errorObj.message, { ...errorObj, data: {}, error: error });
        throw Error(errorObj.message);
    }
}

function addNewUser(newUser: NewUser): Promise<boolean> {
    const url = "/api/users";

    return new Promise((resolve: (object: boolean) => void) => {

        if (!newUser.password) {
            resolve(false);
            return;
        }

        const formData = new FormData();
        formData.append("name", newUser.name);
        formData.append("password", newUser.password);
        formData.append("role", newUser.role);

        requestPromiseJson("POST", url, formData).then(([status]) => {
            if (status === 200 || status === 201) {
                resolve(true);
            } else {
                resolve(false);
            }
        }).catch(() => {
            resolve(false);
        });
    });
}

function deleteUser(username: string): Promise<boolean> {
    const url = "/api/users";

    const headers = {
        "user": username
    };

    return new Promise((resolve: (object: boolean) => void) => {

        requestPromiseJson("DELETE", url, null, headers).then(([status]) => {
            if (status === 204) {
                resolve(true);
            } else {
                resolve(false);
            }
        }).catch(() => {
            resolve(false);
        });
    });
}

// BLAIS5-3816
async function patchUserRolesAndPermissions(user: string, role: string, currentServerParks: string[], currentDefaultServerPark: string): Promise<PatchUserRoleResponse> {
    try {
        const authManager = new AuthManager();
        const url = `/api/users/${user}/rolesAndPermissions`;
        const headers = authManager.authHeader();
        const data = { role, currentServerParks, currentDefaultServerPark };

        const res = await axios.patch(url, data, { headers });

        return {
            status: res.status,
            message: res.data.message
        };
    } catch (error) {
        const errorObj = handleAxiosError(error);
        console.error(errorObj.message, errorObj);
        return errorObj;
    }
}

export { getAllUsers, getUser, addNewUser, deleteUser, patchUserRolesAndPermissions };
