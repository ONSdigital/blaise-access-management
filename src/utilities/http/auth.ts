import {requestPromiseJson} from "./requestPromise";
import {User} from "../../../Interfaces";

type authLoginUserResponse = [boolean, User | null];

function loginUser(username: string, password: string): Promise<authLoginUserResponse> {
    console.log(`Call to login user with user ${username}`);
    const url = `/api/login/${username}`;

    const headers = {
        "password": password,
    };

    return new Promise((resolve: (object: authLoginUserResponse) => void) => {
        requestPromiseJson("GET", url, null, headers).then(([status, data]) => {
            console.log(`Response from login user: Status ${status}, data ${data}`);
            if (status === 200) {
                resolve([true, data]);
            } else if (status === 403) {
                resolve([true, null]);
            } else {
                resolve([false, null]);
            }
        }).catch((error: Error) => {
            console.error(`Response from login user Failed: Error ${error}`);
            resolve([false, null]);
        });
    });
}

export {loginUser};
