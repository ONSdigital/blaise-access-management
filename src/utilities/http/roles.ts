import {requestPromiseJsonList} from "./requestPromise";
import {Role} from "../../../Interfaces";

type getRolesListResponse = [boolean, Role[]];

function getAllRoles(): Promise<getRolesListResponse> {
    console.log("Call to getAllRoles");
    const url = "/api/roles";

    return new Promise((resolve: (object: getRolesListResponse) => void) => {
        requestPromiseJsonList("GET", url).then(([success, data]) => {
            console.log(`Response from get all roles ${(success ? "successful" : "failed")}, data list length ${data.length}`);
            resolve([success, data]);
        }).catch((error: Error) => {
            console.error(`Response from get all roles Failed: Error ${error}`);
            resolve([false, []]);
        });
    });
}

export {getAllRoles};
