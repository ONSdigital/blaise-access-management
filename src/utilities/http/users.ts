import {requestPromiseJson, requestPromiseJsonList} from "./requestPromise";
import {User} from "../../../Interfaces";

type getUsersListResponse = [boolean, User[]];

function getAllUsers(): Promise<getUsersListResponse> {
    console.log("Call to getAllUsers");
    const url = "/api/users";

    return new Promise((resolve: (object: getUsersListResponse) => void) => {
        requestPromiseJsonList("GET", url).then(([success, data]) => {
            console.log(`Response from get all users ${(success ? "successful" : "failed")}, data list length ${data.length}`);
            resolve([success, data]);
        }).catch((error: Error) => {
            console.error(`Response from get all users Failed: Error ${error}`);
            resolve([false, []]);
        });
    });
}

function addNewUser(newUser: User): Promise<boolean> {
    console.log("Call to addNewUser");

    const url = "/api/users";

    return new Promise((resolve: (object: boolean) => void) => {

        if (newUser.password === undefined) {
            resolve(false);
            return;
        }

        const formData = new FormData();
        formData.append("name", newUser.name);
        formData.append("password", newUser.password);
        formData.append("role", newUser.role);

        requestPromiseJson("POST", url, formData).then(([status, data]) => {
            console.log(`Response from add new user: Status ${status}, data ${data}`);
            if (status === 201) {
                resolve(true);
            } else {
                resolve(false);
            }
        }).catch((error: Error) => {
            console.error(`Response from add new user Failed: Error ${error}`);
            resolve(false);
        });
    });
}

export {getAllUsers, addNewUser};
