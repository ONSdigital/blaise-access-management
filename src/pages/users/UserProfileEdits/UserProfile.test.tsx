import React from "react";
import { render, cleanup, waitFor, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import { MemoryRouter, useParams } from "react-router-dom";
import UserProfile from "./UserProfile";
import * as http from "../../../api/http";

jest.mock("react-router-dom", () => ({
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    ...jest.requireActual("react-router-dom"),
    useParams: jest.fn()
}));

jest.mock("../../../api/http", () => ({
    getUser: jest.fn()
}));

const mockUserDetails = {
    data: {
        name: "testUser",
        role: "IPS Manager",
        defaultServerPark: "gusty",
        serverParks: ["gusty", "cma"]
    },
    status: 200,
    message: "Successfully fetched user details for testUser"
};

const mockState = {
    pathname: `/users/${mockUserDetails.data.name}`,
    state: { currentUser: "currentUser", updatedPanel: null }
};

beforeEach(() => {
    jest.clearAllMocks();
    (useParams as jest.Mock).mockReturnValue({ user: mockUserDetails.data.name });
});

afterEach(() => cleanup());

describe("UserProfile Component", () => {
    it("matches the snapshot", async () => {
        (http.getUser as jest.Mock).mockResolvedValue(mockUserDetails);

        const { asFragment } = render(
            <MemoryRouter initialEntries={[mockState]}>
                <UserProfile />
            </MemoryRouter>
        );

        // Wait for state update
        await act(async () => {});

        await waitFor(() => {
            expect(asFragment()).toMatchSnapshot();
        });
    });

    it("displays user details on successful fetch", async () => {
        (http.getUser as jest.Mock).mockResolvedValue(mockUserDetails);

        const { findByText } = render(
            <MemoryRouter initialEntries={[mockState]}>
                <UserProfile />
            </MemoryRouter>
        );

        expect(await findByText(mockUserDetails.data.name)).toBeVisible();
        expect(await findByText(mockUserDetails.data.role)).toBeVisible();
        expect(await findByText(mockUserDetails.data.defaultServerPark)).toBeVisible();
        expect(await findByText(mockUserDetails.data.serverParks.join(", "))).toBeVisible();
    });

    it("displays error message on fetch failure", async () => {
        (http.getUser as jest.Mock).mockRejectedValue(new Error("Unable to load user details, please try again. If this continues, please the contact service desk."));

        const { findByText } = render(
            <MemoryRouter initialEntries={[mockState]}>
                <UserProfile />
            </MemoryRouter>
        );

        expect(await findByText(/Unable to load user details, please try again. If this continues, please the contact service desk./i)).toBeVisible();
    });
});