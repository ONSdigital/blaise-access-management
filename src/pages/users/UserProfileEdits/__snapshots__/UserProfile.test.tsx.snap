// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile Component matches the snapshot 1`] = `
<DocumentFragment>
  <nav
    aria-label="Breadcrumb"
    class="ons-breadcrumb"
  >
    <ol
      class="ons-breadcrumb__items ons-u-fs-s"
    >
      <li
        class="ons-breadcrumb__item"
        id="breadcrumb-0"
      >
        <a
          class="ons-breadcrumb__link"
          href="/"
        >
          Home
        </a>
        <svg
          class="ons-svg-icon"
          focusable="false"
          viewBox="0 0 8 13"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
            transform="translate(-5.02 -1.59)"
          />
        </svg>
      </li>
      <li
        class="ons-breadcrumb__item"
        id="breadcrumb-1"
      >
        <a
          class="ons-breadcrumb__link"
          href="/users"
        >
          Manage users
        </a>
        <svg
          class="ons-svg-icon"
          focusable="false"
          viewBox="0 0 8 13"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
            transform="translate(-5.02 -1.59)"
          />
        </svg>
      </li>
    </ol>
  </nav>
  <main
    class="ons-page__main ons-u-mt-no"
    id="main-content"
  >
    <h1
      class="ons-u-mb-l"
    >
      User Profile
    </h1>
    <div
      class="ons-summary"
    >
      <div
        class="ons-summary__group"
        id="personal-details"
      >
        <h2
          class="ons-summary__group-title"
        >
          Personal details
        </h2>
        <dl
          class="ons-summary__items"
        >
          <div
            class="ons-summary__item"
            id="name-row"
          >
            <div
              class="ons-summary__row ons-summary__row--has-values"
              id="name"
            >
              <dt
                class="ons-summary__item-title"
              >
                <div
                  class="ons-summary__item--text"
                >
                  Name
                </div>
              </dt>
              <dd
                class="ons-summary__values"
              >
                <span
                  class="ons-summary__text"
                >
                  testUser
                </span>
              </dd>
              <dd
                class="ons-summary__actions"
              />
            </div>
          </div>
          <div
            class="ons-summary__item"
            id="password-row"
          >
            <div
              class="ons-summary__row ons-summary__row--has-values"
              id="password"
            >
              <dt
                class="ons-summary__item-title"
              >
                <div
                  class="ons-summary__item--text"
                >
                  Password
                </div>
              </dt>
              <dd
                class="ons-summary__values"
              >
                <span
                  class="ons-summary__text"
                >
                  •••••••••
                </span>
              </dd>
              <dd
                class="ons-summary__actions"
              >
                <a
                  class="ons-summary__button"
                  href="/users/change-password/testUser"
                >
                  <span
                    aria-hidden="true"
                    class="ons-summary__button-text"
                  >
                    Change
                  </span>
                  <span
                    class="ons-u-vh"
                  >
                    Change password
                  </span>
                </a>
              </dd>
            </div>
          </div>
        </dl>
      </div>
      <div
        class="ons-summary__group"
        id="roles-and-permissions"
      >
        <h2
          class="ons-summary__group-title"
        >
          Roles and permissions
        </h2>
        <div
          class="ons-summary__item"
          id="role-row"
        >
          <div
            class="ons-summary__row ons-summary__row--has-values"
            id="role"
          >
            <dt
              class="ons-summary__item-title"
            >
              <div
                class="ons-summary__item--text"
              >
                Role
              </div>
            </dt>
            <dd
              class="ons-summary__values"
            >
              <span
                class="ons-summary__text"
              >
                IPS Manager
              </span>
            </dd>
            <dd
              class="ons-summary__actions"
            >
              <a
                class="ons-summary__button"
                href="/users/change-role/testUser"
              >
                <span
                  aria-hidden="true"
                  class="ons-summary__button-text"
                >
                  Change
                </span>
                <span
                  class="ons-u-vh"
                >
                  Change role
                </span>
              </a>
            </dd>
          </div>
        </div>
        <div
          class="ons-summary__item"
          id="default-server-park-row"
        >
          <div
            class="ons-summary__row ons-summary__row--has-values"
            id="default-server-park"
          >
            <dt
              class="ons-summary__item-title"
            >
              <div
                class="ons-summary__item--text"
              >
                Default Server Park
              </div>
            </dt>
            <dd
              class="ons-summary__values"
            >
              <span
                class="ons-summary__text"
              >
                gusty
              </span>
            </dd>
            <dd
              class="ons-summary__actions"
            />
          </div>
        </div>
        <div
          class="ons-summary__item"
          id="server-parks-row"
        >
          <div
            class="ons-summary__row ons-summary__row--has-values"
            id="server-parks"
          >
            <dt
              class="ons-summary__item-title"
            >
              <div
                class="ons-summary__item--text"
              >
                Server Parks
              </div>
            </dt>
            <dd
              class="ons-summary__values"
            >
              <span
                class="ons-summary__text"
              >
                gusty, cma
              </span>
            </dd>
            <dd
              class="ons-summary__actions"
            />
          </div>
        </div>
      </div>
      <div
        class="ons-summary__group"
        id="actions"
      >
        <h2
          class="ons-summary__group-title"
        >
          Actions
        </h2>
        <div
          id="actions-row"
        >
          <a
            class="ons-summary__button"
            href="/users/delete/testUser"
          >
            <span
              aria-hidden="true"
              class="ons-summary__button-text"
            >
              Delete
            </span>
            <span
              class="ons-u-vh"
            >
              Delete user
            </span>
          </a>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
